File: backend-java/src/main/java/org/egualpam/contexts/observable/shared/application/domain/AggregateRoot.java
package org.egualpam.contexts.observable.shared.application.domain;

import java.util.HashSet;
import java.util.Set;

public abstract class AggregateRoot {

    private final Set<DomainEvent> domainEvents = new HashSet<>();

    protected void add(DomainEvent event) {
        domainEvents.add(event);
    }

    public Set<DomainEvent> pullDomainEvents() {
        var pulled = Set.copyOf(domainEvents);
        domainEvents.clear();
        return pulled;
    }

    public abstract AggregateId id();

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) return false;
        AggregateRoot that = (AggregateRoot) o;
        return id().equals(that.id());
    }

    @Override
    public int hashCode() {
        return id().hashCode();
    }
}

File: backend-kotlin/src/main/kotlin/org/egualpam/contexts/observable/shared/application/domain/AggregateRoot.kt
package org.egualpam.contexts.observable.shared.application.domain

abstract class AggregateRoot {

  private val domainEvents = mutableSetOf<DomainEvent>()

  protected fun add(domainEvent: DomainEvent) = domainEvents.add(domainEvent)

  fun pullDomainEvents(): Set<DomainEvent> {
    val pulled = domainEvents.toSet()
    domainEvents.clear()
    return pulled
  }

  abstract fun getId(): AggregateId

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as AggregateRoot

    return getId() == other.getId()
  }

  override fun hashCode(): Int {
    return getId().hashCode()
  }
}
